internal/server/data/grant.go

```diff
@@ -195,6 +217,106 @@ func ListGrants(tx ReadTxn, opts ListGrantsOptions) ([]models.Grant, error) {
 	return result, rows.Err()
 }
 
+func grantsByDestination(query *querybuilder.Query, destination string) {
+	query.B("AND (resource = ? OR resource LIKE ?)", destination, destination+".%")
+}
+
+type GrantsMaxUpdateIndexOptions struct {
+	ByDestination string
+}
+
+// GrantsMaxUpdateIndex returns the maximum update_index all the grants that
+// match the query. This MUST include soft-deleted rows as well.
+//
+// TODO: any way to assert this tx has the right isolation level?
+func GrantsMaxUpdateIndex(tx ReadTxn, opts GrantsMaxUpdateIndexOptions) (int64, error) {
+	query := querybuilder.New("SELECT max(update_index) FROM grants")
+	query.B("WHERE organization_id = ?", tx.OrganizationID())
+
+	if opts.ByDestination != "" {
+		grantsByDestination(query, opts.ByDestination)
+	}
+
+	var result int64
+	err := tx.QueryRow(query.String(), query.Args...).Scan(&result)
+	return result, err
+}
+
+type Listener struct {
+	sqlDB   *sql.DB
+	pgxConn *pgx.Conn
+}
+
+// WaitForNotification blocks until the listener receivers a notification on
+// one of the channels, or until the context is cancelled.
+// Returns the notification on success, or an error on failure or timeout.
+func (l *Listener) WaitForNotification(ctx context.Context) (*pgconn.Notification, error) {
+	return l.pgxConn.WaitForNotification(ctx)
+}
+
+func (l *Listener) Release(ctx context.Context) error {
+	var errs []error
+	if _, err := l.pgxConn.Exec(ctx, `UNLISTEN *`); err != nil {
+		errs = append(errs, err)
+	}
+	if err := pgxstdlib.ReleaseConn(l.sqlDB, l.pgxConn); err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to unlisten to postgres channels: %v", errs)
+	}
+	return nil
+}
+
+type ListenForGrantsOptions struct {
+	OrgID         uid.ID
+	ByDestination string
+}
+
+// ListenForGrantsNotify starts listening for notification on one or more
+// postgres channels for notifications that a grant has changed. The channels to
+// listen on are determined by opts. Use Listener.WaitForNotification to block
+// and receive notifications.
+//
+// If error is nil the caller must call Listener.Release to return the database
```

Comment:
I think so ya. I think the `UNLISTEN` shouldn't error, but the call to release will try to close a DB connection that is already closed, so that might error.